OHOS-Media/fordham-ai-career-toolbox
# feat: Refactor all components and add custom hooks

# Authentication System Overhaul and Application Restructuring

## Overview
This PR introduces a major overhaul of our authentication system and restructures our application to improve modularity, maintainability, and user experience. We've implemented a centralized authentication context, created new components, and reorganized our existing code to better separate concerns.

## Key Changes

### Authentication
1. Implemented `AuthContext` and `AuthProvider` for centralized auth state management.
2. Created a custom `useAuth` hook for easy access to auth functionality throughout the app.
3. Implemented `useApi` hook for standardized API calls with built-in error handling and loading states.
4. Updated login flow to use a popup window for Google OAuth.
5. Improved error handling for failed login attempts, especially for non-Fordham email addresses.

### Error Handling
1. Modified the existing `ErrorAlert` component to work with the new authentication flow.
2. Implemented a centralized error display mechanism in the `AuthContext`.
3. Added automatic error message dismissal after 5 seconds and manual dismissal option.

### Component Structure
1. Created a new `Hero` component for the landing page.
2. Developed a new `Toolbox` page to house the keyword extractor functionality.
3. Implemented a `Layout` component to maintain consistent structure across pages.
4. Updated the `Nav` component to conditionally render login/user info based on auth state.
5. Enhanced the `Button` component to be more flexible and reusable.
7. Fixed issues with the `Logo` component to properly display SVG in Next.js environment.

### Page Updates
1. Simplified `index.js` (Home page) to use the new `Hero` component.
2. Created `toolbox.jsx` page for the keyword extractor tool.
3. Updated `profile.js` to use the new auth context and layout.

### Server-side Changes
1. Updated Google OAuth callback route to handle both successful and failed login attempts.
2. Implemented a new route for failed logins that communicates the error back to the client.
3. Refined the logout route for better error handling.

## Detailed Changes

### New Files
- `context/AuthContext.js`: Implements AuthContext and AuthProvider.
- `hooks/useApi.js`: Custom hook for API calls.
- `components/Hero.jsx`: New Hero component for the landing page.
- `pages/toolbox.jsx`: New page for the keyword extractor tool.
- `components/Layout.jsx`: New Layout component for consistent page structure.

### Updated Files
- `components/Nav/Nav.jsx`: Updated to use AuthContext and simplified login/logout flow.
- `components/Button.jsx`: Enhanced with onClick prop and additional customization options.
- `components/Logo.jsx`: Fixed to properly display SVG using Next.js Image component.
- `pages/index.js`: Simplified to use new Hero component.
- `pages/profile.js`: Updated to use AuthContext and new Layout component.
- `pages/_app.js`: Wrapped with AuthProvider for global auth state.

### Removed Files
- `components/LoginPopup.jsx`: Functionality integrated into AuthContext.

## Testing Instructions
1. Test the login flow by clicking the "Log In" button in the navigation bar.
2. After logging in, verify that the navigation bar updates to show the user's name.
4. Navigate to the profile page and test the logout functionality.
5. Test the keyword extractor tool on the new Toolbox page.
6. Verify that the logo displays correctly on all pages.

## Notes
- This PR introduces breaking changes to the authentication flow. Please review and update your local environments accordingly.
- The `NEXT_PUBLIC_SERVER_URL` environment variable must be set for the authentication to work correctly.
- SVG handling may require additional configuration in deployment environments.
